Add Guardian tutorial (#1203)
viamrobotics/docs: docs/tutorials/projects/guardian.md

---
title: "Security Guardian"
linkTitle: "Security Guardian"
weight: 50
type: "docs"

> description: "Make a functional guardian with a servo motor, some LEDs, a camera, and the ML and Vision Service to detect people and pets."

Reviewer: ```suggestion
description: "Make a functional guardian with a servo motor, some LEDs, a camera, and the ML Model and Vision Service to detect people and pets."
```

webmSrc: "/tutorials/videos/foam-dart.webm"
mp4Src: "/tutorials/videos/foam-dart.mp4"
videoAlt: "A guardian detecting a person or pet."
images: ["/tutorials/img/guardian/preview.gif"]
tags: ["camera", "vision", "detector", "python"]
no_list: true
# Author: Naomi Pentrel
---

> In the run up to the new Zelda release, I realized you could build a stationary guardian robot with a servo and a camera.

Author: ```suggestion
In the run up to the new Zelda release, I realized you can build a stationary guardian robot with a servo and a camera.
```

> Adding a bit of [machine learning](/services/ml/), you could then make the guardian detect objects or people or pets and follow them around by rotating its head.

Author: ```suggestion
Adding a bit of [machine learning](/services/ml/), you can then make the guardian detect objects or people or pets and follow them around by rotating its head.
```

Luckily, I am not the first one to have the idea to build a guardian and there was already a [brilliant guardian 3D model](https://www.thingiverse.com/thing:2391826) on Thingiverse with space for LEDs and a servo.

> In this tutorial, I will walk you through the steps to build your own functional guardian.

Author: ```suggestion
In this tutorial, I will walk you through the steps to build your own functional guardian with a [servo](/components/servo), a [camera](/components/camera), some LEDs and the [ML Model service](/services/ml) and [Vision Service](/services/vision).
```

> 

Author: ```suggestion
Here's a video of the finished guardian detecting me:

```

> {{<video webm_src="../../img/guardian/preview.webm" mp4_src="../../img/guardian/preview.mp4" alt="Guardian robot detects person and rotates head to follow them around">}}

Reviewer: This video is great-- opt. you could probably add more in intro to introduce video & "sell the project", being like "I built this with xyz components and Viam ML model service etc." rather than just "I realized I could build something like this"!

- **acrylic paint**: I ordered armour modelling paint but found that mixing my own colors from a regular acrylic paint set worked best for me.
- **modeling grass, stones, glue**: The Army Painter makes a Battlefields Basing Set which comes with all of this.
- **a base for the guardian**: I used a wooden disk with a hole cut in the middle and a box with a hole in the top underneath.
    <div style="max-width: 400px">
    <img src="../../img/guardian/base.jpg" alt="Wooden guardian base">
    </div>
- **ground texture**: If you want the base to look more natural, you can use Vallejo Ground Texture Acrylic or something similar to create patches that look like stone.
- **wire**: To allow you to position the legs better, you can thread wire through them.

## Software requirements

You will use the following software in this tutorial:

- [Python 3](https://www.python.org/downloads/)
- [`viam-server`](/installation#install-viam-server): Follow the [installation instructions](/installation#install-viam-server) to install `viam-server` on your Raspberry Pi.

## Assemble the robot

You can view a time lapse of the robot assembly here:

> {{<video webm_src="../../img/guardian/timelapse.webm" mp4_src="../../img/guardian/timelapse.mp4" alt="Timelapse of guardian assembly">}}

Reviewer: Love this video!

> You can view a time lapse of the robot assembly here:

Author: ```suggestion
You can view a timelapse of the robot assembly here:
```

Then put both parts of the head together.

Your servo probably came with mounting screws and a plastic horn for the gear.
Use the screws to attach the horn to the base of the head.

Next, get your Raspberry Pi and your servo and connect the servo to the Raspberry Pi by connecting the PWM wire to pin 12, the power wire to pin 2, and the ground wire to pin 8.
Then attach the head to the servo.

![A Raspberry Pi connected to a FS90R servo. The yellow PWM wire is attached to pin twelve on the raspberry pi. The red five-volt wire is attached to pin two. The black ground wire is attached to pin eight](/tutorials/img/single-component-tutorials-servo-mousemover/servo-wiring-diagram.png)

Next, get the three 10mm RGB LEDs ready.
Attach the common cathode of each LED to a [ground pin](https://pinout.xyz/pinout/ground) on your Raspberry Pi.
Attach the wires for the red and the blue LEDs to [GPIO pins](https://pinout.xyz/pinout/wiringpi).

<div class="td-max-width-on-larger-screens" style="margin: auto">
  <img src="../../img/guardian/assembled-testing.jpg" alt="Components assembled for testing" style="max-width: 400px;" />
</div>

> Before continuing with assembly, you should test your components work as expected.

Author: ```suggestion
Before continuing with assembly, test your components work as expected.
```

Go to the **Setup** tab of your new robot's page and follow the steps [to install `viam-server` on your computer](/installation).

### Configure the components

{{< tabs >}}
{{% tab name="Builder UI" %}}

Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
Click on the **Components** subtab and navigate to the **Create component** menu.

1. **Add the board.**

    Enter `local` for the name for your [board component](/components/board/), select the type `board`, and select the `pi` model.
    Then click **Create component**.

2. **Add the camera.**

    Create a [camera component](/components/camera/) with the name `cam`, the type `camera` and the model `webcam`.
    Click **Create Component** to add the camera.
    In the new camera panel, click the **Video Path** field to reveal a drop-down populated with camera paths that have been identified on your machine.
    Select `mmal service 16.1 (platform:bcm2835_v4l2-0)`.

3. **Add the servo.**

    Create a [servo component](/components/servo/) with the name `servo`, the type `servo` and the model `pi`.
    Click **Create Component** to add the servo.

>     Configure the attributes, by adding the name of your board `local` and the pin you connected your servo to `12`:

Reviewer: ```suggestion
    Configure the attributes by adding the name of your board, `local`, and the {{< glossary_tooltip term_id="pin-number" text="pin number" >}} of the pin on `local` that you connected your servo PWM wire to, `12`:
```

    ```json
    {
        "pin": "12",
        "board": "local"
    }
    ```

Click **Save config** in the bottom left corner of the screen.

{{% /tab %}}
{{% tab name="Raw JSON" %}}

> On the [`Raw JSON` tab](/manage/configuration/#the-config-tab), replace the configuration with the following configuration which configures your board, your camera, and your servo on pin `12`:

Reviewer: ```suggestion
On the [`Raw JSON` tab](/manage/configuration/#the-config-tab), replace the configuration with the following JSON configuration for your board, your camera, and your servo with its PWM wire wired to  {{< glossary_tooltip term_id="pin-number" text="pin number" >}} `12`:
```
Or something like that! I think point out where the pin number comes from helps a bit as well as the tooltip

   Then click **Create Service**.

   In the new ML Model panel, select **Path to Existing Model On Robot** for the **Deployment**.

   Then specify the absolute **Model Path** as `/home/pi/effdet0.tflite` and the **Label Path** as `/home/pi/labels.txt`.

2. **Add a vision service.**

   Next, add a [detector](/services/vision/detection) as a vision service to be able to make use of the ML model.
   Create an vision service with the name `detector`, the type `vision` and the model `mlmodel`.
   Then click **Create Service**.

   In the new detector panel, select the `mlmodel` you configured in the previous step.

   Click **Save config** in the bottom left corner of the screen.

3. **Add a `transform` camera.**

   To be able to test that the vision service is working, add a `transform` camera which will add bounding boxes and labels around the objects the service detects.

   Click on the **Components** subtab and navigate to the **Create component** menu.
   Create a [transform camera](/components/camera/transform) with the name `transform_cam`, the type `camera` and the model `transform`.

>    Replace the attributes JSON object with the following:

Reviewer: If you have time here a bit more explanation of what's happening with the pipeline attributes and transform config could be helpful-- I was initially confused as to why attributes were being replaced with attributes, explaining that it's a "transformation object" could be useful!

>    Replace the attributes JSON object with the following:

Author: ```suggestion
   Replace the attributes JSON object with the following object which specifies the camera source that the `transform` camera will be using and defines a pipeline that adds the defined `detector`:
```

        "local"
      ]
    },
    {
      "name": "transform_cam",
      "type": "camera",
      "model": "transform",
      "attributes": {
        "source": "cam",
        "pipeline": [
          {
            "attributes": {
              "detector_name": "detector",
              "confidence_threshold": 0.6
            },
            "type": "detections"
          }
        ]
      },
      "depends_on": []
    }
  ],
  "services": [
    {
      "name": "mlmodel",
      "type": "mlmodel",
      "model": "tflite_cpu",
      "attributes": {
        "model_path": "/home/pi/effdet0.tflite",
        "label_path": "/home/pi/labels.txt",
        "num_threads": 1
      }
    },
    {
      "name": "detector",
      "type": "vision",
      "model": "mlmodel",
      "attributes": {
        "mlmodel_name": "mlmodel"
      }
    }
  ]
}
```

Click **Save config** in the bottom left corner of the screen.

{{% /tab %}}
{{< /tabs >}}

Navigate to your [robot's Control tab](/manage/fleet/robots/#control) to test the transform camera.
Click on the transform camera panel and toggle the camera on, then point your camera at a person or pet to test if the vision service detects them.

> You should see a bounding box with label around different objects.

Reviewer: ```suggestion
You should see a bounding box with labels around different objects.
```

<div style="max-width: 600px" >
    <img src="../../img/guardian/test-transform.jpg" alt="the control tab transform camera panel">
</div>

## Program the Guardian

> With the guardian completely configured and the configuration tested, it's time to program the guardian to behave like a real guardian.

Reviewer: And do what?

> With the guardian completely configured and the configuration tested, it's time to program the guardian to behave like a real guardian.

Reviewer: ```suggestion
With the guardian completely configured and the configuration tested, it's time to make the robot guardian behave like a "real" guardian.
Next, you'll use the [Viam Python SDK](/program/sdks) to program it to detect people and pets and scare them with music and LEDs. 
```
MAybe something like this-- think it would at least be helpful to spec that you're coding in Python and using an SDK to do so!

Author: With the guardian completely configured and the configuration tested, it's time to make the robot guardian behave like a "real" guardian by programming the person and pet detection, the lights, music, and the movement:

> With the guardian completely configured and the configuration tested, it's time to program the guardian to behave like a real guardian.

Author: ```suggestion
With the guardian completely configured and the configuration tested, it's time to make the robot guardian behave like a "real" guardian by programming the person and pet detection, the lights, music, and the movement:
```

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
python3 -m pip install --user virtualenv
python3 -m venv env
source env/bin/activate
```

Now, install the Python Viam SDK and the VLC module:

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
pip3 install viam-sdk python-vlc
```

### Connect

Next, go to the **Code Sample** tab on your robot page and select **Python**.

Copy the boilerplate code.
This code snippet imports all the necessary packages and sets up a connection with the Viam app in the cloud.

Next, create a file named <file>main.py</file> and paste the boilerplate code from the **Code Sample** tab of the Viam app into your file.
Then, save your file.

Run the code to verify that the Viam SDK is properly installed and that the `viam-server` instance on your robot is live.

You can run your code by typing the following into your terminal:

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
python3 main.py
```

The program prints a list of robot resources.

On top of the packages that the code sample snippet imports, add the `random` and the `vlc` package to the imports.

> The top of your code now look like this:

Reviewer: ```suggestion
The top of your code should now look like this:
```

async def main():
    robot = await connect()
    local = Board.from_robot(robot, 'local')
    cam = Camera.from_robot(robot, "cam")
    img = await cam.get_image()
    servo = Servo.from_robot(robot, "servo")
    red_leds = LedGroup([
        await local.gpio_pin_by_name('22'),
        await local.gpio_pin_by_name('24'),
        await local.gpio_pin_by_name('26')
    ])
    blue_leds = LedGroup([
        await local.gpio_pin_by_name('11'),
        await local.gpio_pin_by_name('13'),
        await local.gpio_pin_by_name('15')
    ])

    await blue_leds.led_state(True)

    music_player = vlc.MediaPlayer("guardian.mp3")

    # grab Viam's vision service for the detector
    detector = VisionClient.from_robot(robot, "detector")
    while True:
        # move head periodically left and right until movement is spotted.
        living_creature = await idle_and_check_for_living_creatures(cam, detector, servo, blue_leds, red_leds, music_player)
        await focus_on_creature(living_creature, img.width, servo)
    # Don't forget to close the robot when you're done!
    await robot.close()

if __name__ == '__main__':
    asyncio.run(main())
```

> Now, run code:

Author: ```suggestion
Now, run the code:
```

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
python3 main.py
```

> 

Author: ```suggestion

If everything works, your guardian should now start to idle and when it detects humans or dogs or cats turn red, start music, and focus on the detected being:

```

              "detector_name": "detector",
              "confidence_threshold": 0.6
            },
            "type": "detections"
          }
        ]
      },
      "depends_on": []
    }
  ],
  "services": [
    {
      "name": "mlmodel",
      "type": "mlmodel",
      "model": "tflite_cpu",
      "attributes": {
        "model_path": "/home/pi/effdet0.tflite",
        "label_path": "/home/pi/labels.txt",
        "num_threads": 1
      }
    },
    {
      "name": "detector",
      "type": "vision",
      "model": "mlmodel",
      "attributes": {
        "mlmodel_name": "mlmodel"
      }
    }
  ]
}
```

Click **Save config** in the bottom left corner of the screen.

{{% /tab %}}
{{< /tabs >}}

Navigate to your [robot's Control tab](/manage/fleet/robots/#control) to test the transform camera.
Click on the transform camera panel and toggle the camera on, then point your camera at a person or pet to test if the vision service detects them.
You should see bounding boxes with labels around different objects.

<div style="max-width: 600px" >
    <img src="../../img/guardian/test-transform.jpg" alt="the control tab transform camera panel">
</div>

## Program the Guardian

> With the guardian completely configured and the configuration tested, it's time to make the robot guardian behave like a "real" guardian by programming the person and pet detection, the lights, music, and the movement:

Reviewer: ```suggestion
With the guardian completely configured and the configuration tested, it's time to make the robot guardian behave like a "real" guardian by programming the person and pet detection, lights, music, and movement.
```

> The following sections walk you through each part of the code.

Reviewer: ```suggestion
```

        await local.gpio_pin_by_name('13'),
        await local.gpio_pin_by_name('15')
    ])

    await blue_leds.led_state(True)

    music_player = vlc.MediaPlayer("guardian.mp3")

    # grab Viam's vision service for the detector
    detector = VisionClient.from_robot(robot, "detector")
    while True:
        # move head periodically left and right until movement is spotted.
        living_creature = await idle_and_check_for_living_creatures(cam, detector, servo, blue_leds, red_leds, music_player)
        await focus_on_creature(living_creature, img.width, servo)
    # Don't forget to close the robot when you're done!
    await robot.close()

if __name__ == '__main__':
    asyncio.run(main())
```

Now, run the code:

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
python3 main.py
```

If everything works, your guardian should now start to idle and when it detects humans or dogs or cats turn red, start music, and focus on the detected being:

{{<video webm_src="../../img/guardian/guardian-finished.webm" mp4_src="../../img/guardian/guardian-finished.mp4" alt="FInished guardian">}}

## Run the program automatically

One more thing.

> Right now, you have to run the code manually every time you want your guardian to do get to work.

Reviewer: ```suggestion
Right now, you have to run the code manually every time you want your guardian to work.
```

You can also configure Viam to automatically run your code as a [process](/manage/configuration/#processes).

> To be able to run the python script from your Raspberry Pi, you need install the Python SDK on your Raspberry Pi and copy your code onto the Raspberry Pi.

Reviewer: ```suggestion
To be able to run the python script from your Raspberry Pi, you need to install the Python SDK on your Raspberry Pi and copy your code onto the Raspberry Pi.
```
